### DISCUSSION TITLE
Send data (8x8 uint16_t array) from CrazyFlie app on STM32 to AI deck app

### DISCUSSION THREAD

**Discussion by kkalenbe** [bitcraze/discussions#572] [Q&A]
Hello everyone,

I am currently trying to send data (8x8 uint16_t array) from a custom CrazyFlie app on STM32 to a custom app on the AI deck. I am having trouble in finding out how to do this as I did not find any examples for this communication online.

I saw this thread, but this did not really help me much. https://github.com/orgs/bitcraze/discussions/519

Could someone point me in the right direction or even provide a short example if there is one around?

Thank you very much for the support and I am looking forward to hear from you :)

Best,

Konstantin

**Comment by evoggy** [comment]
Hi!

Did you have a look at the examples in the thread you linked to? It sounds like CPX could be used for this, sending the data from the STM32 to GAP8 via the ESP32, but unfortunately there's currently no example showing this. If sending via the ESP32 has too high latency for the application it's also possible to use the UART directly between the GAP8 and STM32.

**Reply by kkalenbe** [reply]
Hey @evoggy and thank you very much for your fast reply. I have seen the example but I am a bit confused as the `cpx.h` file has changed since this example has been published. The function `cpxSendPacketBlocking` and many more do not exist in the `cpx.h` anymore. Should I use the `cpx_internal_router.h` file for my usecase? And in addition, what is the `CPXFunction_t` I should use for sending from STM32 to the AI deck?

Thanks again for your help! :)

**Reply by kkalenbe** [reply]
Hey @evoggy. When I am trying to  use the CPX functions provided (https://github.com/bitcraze/aideck-gap8-examples/tree/master/lib/cpx) in the into my AI deck app, I run into this error, which occurs when using the `cpxInit()` function on the AI deck. Do you have any idea on how to solve this? 

Thank you very much!!

```
/home/konstantin/master_thesis/crazyflie_code/aideck_app/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/lib/cpx/src/com.o: In function `vDataReadyISR':
/home/konstantin/gap_sdk/rtos/freeRTOS/vendors/gwt/pmsis/include/pmsis/backend/pmsis_backend_native_task_api.h:26: undefined reference to `xTimerPendFunctionCallFromISR'
```

**Reply by hmllr** [reply]
I know that you found a workaround for this - but if you would still want to fix the cpx, I think adding 
`APP_CFLAGS += -DconfigUSE_TIMERS=1 -DINCLUDE_xTimerPendFunctionCall=1`
in the Makefile should do the trick (I did not yet test it myself though).

**Comment by RavenLite** [comment]
Hi!

I also have the same requirement to send/receive from GAP8/STM32, I think it will be a simple example to show the most basic function of CPX -- communication.
So, it's there any plan to provide such an example? I guess it will be a great example for the freshman to use AI-deck and CPX.

Best,
Raven

**Reply by kkalenbe** [reply]
Hey @RavenLite,

I got something working without CPX only over UART1. The problem is that after the sending function I need to add a large artificial delay so the whole 128bytes of my array arrive correctly.

Maybe this can help you a bit or you can see why I have to add this delay to get a correct send of the results.

Have a great day,

Konstantin

**Crazyflie app:**

```
int16_t tof_test_uart[] = {1, 0, 0, 0, 0, 0, 0, 2,
                           0, 0, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0,
                           3, 0, 0, 0, 0, 0, 0, 4};
uart1Init(9600);

// Loop
uart1SendData(8 * 8 * sizeof(uint16_t), tof_test_uart);
vTaskDelay(1000);  // The delay I talked about
```

**AI-Deck app:**
```
L2_MEM uint16_t tof_receive_buffer_mm[8*8];  // Define global

pi_uart_conf_init(&uart_conf);
uart_conf.enable_tx = 1;
uart_conf.enable_rx = 1;
uart_conf.baudrate_bps = 9600;
pi_open_from_conf(&uart, &uart_conf);
if (pi_uart_open(&uart))
{
    printf("Uart open failed !\n");
    pmsis_exit(-1);
}

// Loop
pi_uart_read(&uart, tof_receive_buffer_mm, 8 * 8 * sizeof(uint16_t));
```

**Reply by knmcguire** [reply]
If you have gotten it working more reliable, this would make a great example in both the Crazyflie and aideck example repos. 

Unfortunately Marcus is a bit busy to help out with cpx related issues... is there anything that is blocking you at the moment let us know?

**Reply by knmcguire** [reply]
I made this feature ticket on the aideck repo: https://github.com/bitcraze/aideck-gap8-examples/issues/116. For things specific to adding this example to the repo, we can communicate there: https://github.com/bitcraze/aideck-gap8-examples/issues/116

**Reply by kkalenbe** [reply]
> If you have gotten it working more reliable, this would make a great example in both the Crazyflie and aideck example repos.
> 
> Unfortunately Marcus is a bit busy to help out with cpx related issues... is there anything that is blocking you at the moment let us know?

Hello @knmcguire, thanks for opening the feature ticket! If I get the normal UART communication running as desired I will give an update on this thread, providing an example.

**Reply by RavenLite** [reply]
> Hello @knmcguire, thanks for opening the feature ticket! If I get the normal UART communication running as desired I will give an update on this thread, providing an example.

Thanks a lot! I will try to use UART directly.  
BTW, the over-air flashing based on CPX is really awesome to flash gap8 application. My macOS often meets something error while using wired JTAG. I like over-air flashing very much!

**Reply by kkalenbe** [reply]
> Hey @RavenLite,
> 
> I got something working without CPX only over UART1. The problem is that after the sending function I need to add a large artificial delay so the whole 128bytes of my array arrive correctly.
> 
> Maybe this can help you a bit or you can see why I have to add this delay to get a correct send of the results.
> 
> Have a great day,
> 
> Konstantin
> 
> **Crazyflie app:**
> 
> ```
> int16_t tof_test_uart[] = {1, 0, 0, 0, 0, 0, 0, 2,
>                            0, 0, 0, 0, 0, 0, 0, 0,
>                            0, 0, 0, 0, 0, 0, 0, 0,
>                            0, 0, 0, 0, 0, 0, 0, 0,
>                            0, 0, 0, 0, 0, 0, 0, 0,
>                            0, 0, 0, 0, 0, 0, 0, 0,
>                            0, 0, 0, 0, 0, 0, 0, 0,
>                            3, 0, 0, 0, 0, 0, 0, 4};
> uart1Init(9600);
> 
> // Loop
> uart1SendData(8 * 8 * sizeof(uint16_t), tof_test_uart);
> vTaskDelay(1000);  // The delay I talked about
> ```
> 
> **AI-Deck app:**
> 
> ```
> L2_MEM uint16_t tof_receive_buffer_mm[8*8];  // Define global
> 
> pi_uart_conf_init(&uart_conf);
> uart_conf.enable_tx = 1;
> uart_conf.enable_rx = 1;
> uart_conf.baudrate_bps = 9600;
> pi_open_from_conf(&uart, &uart_conf);
> if (pi_uart_open(&uart))
> {
>     printf("Uart open failed !\n");
>     pmsis_exit(-1);
> }
> 
> // Loop
> pi_uart_read(&uart, tof_receive_buffer_mm, 8 * 8 * sizeof(uint16_t));
> ```

Concerning this example and the required delay. I figured that I can repeatedly read from the UART on the AI deck if I close the open and close the UART in every iteration of the loop, and not just in the beginning and end of my programm. Maybe this helps someone else as well.

**Comment by matthewoots** [comment]
I've created a CPX serialization example, for different data types for now it's uint8 struct(uint16,uint16) and 3x3 float matrix to simulate April tag detection results. You can check the PR here https://github.com/bitcraze/aideck-gap8-examples/pull/121
